src/main/scala/io/iohk/ethereum/ledger/BlockImport.scala

```diff
@@ -129,8 +132,18 @@ class BlockImport(
       )
   }
 
-  private def isBetterBranch(block: Block, bestBlock: Block, newTd: BigInt, currentTd: BigInt): Boolean =
-    newTd > currentTd || (blockchainConfig.gasTieBreaker && newTd == currentTd && block.header.gasUsed > bestBlock.header.gasUsed)
+  private def isBetterBranch(block: Block, bestBlock: Block, newTd: BigInt, currentTd: BigInt): Boolean = {
+    lazy val betterTd = newTd > currentTd
+    lazy val tieBreaker =
+      blockchainConfig.gasTieBreaker && newTd == currentTd && block.header.gasUsed > bestBlock.header.gasUsed
+
+    (block.hasCheckpoint, bestBlock.hasCheckpoint) match {
```

Comment:
You are right! Let's consider 2 branches:
- B0 -> C -> B1
- B0 -> B2 -> B3

If we're considering B3 for import and it has a greater TD than B1, then checkpoint C would be reverted by a regular block.

This simple method cannot be adapted to handling checkpoints. What I think we could do is obtain the branch for the newly enqueued block (`BlockQueue#getBranch`: [here](https://github.com/input-output-hk/mantis/blob/f0416e09f629d15fb5454ba7a8cfaaeba82a422c/src/main/scala/io/iohk/ethereum/ledger/BlockImport.scala#L124)) and then compare it using `BranchResolution#resolveBranch`, which as a bonus DRY the code.

WDYT?

Reply:
It would be great to remove this code duplication, my only comments are:
- I feel like we are doing something unnecesary by having both the importing and resolving process do this sort of comparisons. However at this stage is probably something we'll have to live with
- For comparison shouldn't we use `BranchResolution#compareByCheckpoints` instead? There a several thing surrounding BranchResolution#resolveBranch that are probably not necessary

But the above are just technical questions, totally agree with re-using code here!

Reply:
> However at this stage is probably something we'll have to live with

The only way I see around is to take advantage of the improvements in our testnet

> There a several thing surrounding BranchResolution#resolveBranch that are maybe not necessary

Not that much really:
- checking the parent
- checking that headers from queue form a chain

Avoiding that shouldn't be worth refactoring

Reply:
>The only way I see around is to take advantage of the improvements in our testnet

What improvements do you mean?

Btw yesterday I thought of an alternative to your idea, which is closer to what is currently in place on the BlockQueue; if our leaves would be as `Leaf(hash: ByteString, totalDifficulty: BigInt, lastCheckpointBlockNumber)` can't this function be adapted to compare those `lastCheckpointBlockNumber`? Obviously, if we decide that mantis doesn't allow reverts then that information is not enough... but maybe we could rely on OBFT being the one that doesn't produce checkpoints of forks